### PR TITLE
Refactor and cleanup: Jibri and ChatRoomMemberRole

### PR DESCRIPTION
**Pull request by bgrozev**

### CODE REVIEW

File: src/main/java/org/jitsi/jicofo/jibri/JibriSession.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ static private boolean isStartingStatus(JibriIq.Status status)
     /**
      * The owner which will be notified about status changes of this session.
```
Comment (JonathanLennox) [inline src/main/java/org/jitsi/jicofo/jibri/JibriSession.java:98]: Update comment?

File: src/main/java/org/jitsi/jicofo/jibri/JibriSession.java [hunk 2]
```diff
@@ -172,7 +172,7 @@ static private boolean isStartingStatus(JibriIq.Status status)
 
     /**
      * Creates new {@link JibriSession} instance.
-     * @param owner the session owner which will be notified about this session
+     * @param stateListener the session owner which will be notified about this session
```
Comment (JonathanLennox) [inline src/main/java/org/jitsi/jicofo/jibri/JibriSession.java:175]: Update comment?

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomLocalUserRoleChangeEvent.java [hunk 3]
```diff
@@ -17,12 +17,14 @@
...
 
+import org.jitsi.jicofo.xmpp.muc.*;
+
 /**
  * Dispatched to notify interested parties that a change in our role in the
  * source chat room has occurred. Changes may include us being granted admin
  * permissions, or other permissions.
  *
- * @see ChatRoomMemberRole
+ * @see emberRole
```
Comment (JonathanLennox) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomLocalUserRoleChangeEvent.java:27]: Typo

File: src/main/java/org/jitsi/jicofo/ChatRoomRoleAndPresence.java [hunk 4]
```diff
@@ -209,7 +209,7 @@ private void electNewOwner()
             {
                 continue;
             }
-            else if (MemberRole.OWNER.compareTo(member.getRole()) >=0)
+            else if (member.getRole() == MemberRole.OWNER)
```
Comment (JonathanLennox) [inline src/main/java/org/jitsi/jicofo/ChatRoomRoleAndPresence.java:212]: This assumes owner is the highest role?  Would it be more symmetric to have a `hasOwnerRights()` method?

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatMemberImpl.java [hunk 5]
```diff
@@ -229,7 +229,8 @@ void processPresence(Presence presence)
         {
             boolean[] startMuted = { ext.getAudioMuted(), ext.getVideoMuted() };
 
-            if (getRole().compareTo(MemberRole.MODERATOR) < 0)
+            // XXX Is this intended to be allowed for moderators or not?
```
Comment (JonathanLennox) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatMemberImpl.java:232]: It was moderator before, yes? Do you think this is wrong?

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt [hunk 6]
```diff
@@ -65,15 +66,19 @@ abstract class BaseJibri internal constructor(
         TaskPools.ioPool
     )
 
+    private val clientIqHandler: JibriSessionIqHandler = object : JibriSessionIqHandler {
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:69]: Why did you do it this way rather than having `BaseJibri` implement `JibriSessionIqHandler`?

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+     * The current recording session or <tt>null</tt>.
+     */
+    private var jibriSession: JibriSession? = null
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun dispose() {
+        jibriSession?.stop(null)
+        super.dispose()
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt:67]: The Java code nulled `jibriSession` here, presumably to prevent stopping the session twice - why did you remove it?

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+     */
+    override fun handleStartRequest(iq: JibriIq): IQ {
+        val streamIdIsEmpty = StringUtils.isBlank(iq.streamId)
+        return if ((
+            !streamIdIsEmpty &&
+                iq.recordingMode == RecordingMode.STREAM ||
+                iq.recordingMode == RecordingMode.UNDEFINED
+            ) ||
+            streamIdIsEmpty && iq.recordingMode == RecordingMode.FILE
+        ) {
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt:104]: Ugh, is this really the format that ktlint insists on? Any way it can be closed up at all?

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                    is AllBusy -> {
+                        logger.info("Failed to start a Jibri session, all Jibris were busy")
+                        error(iq, XMPPError.Condition.resource_constraint, "all Jibris are busy")
+                    }
+                    is NotAvailable -> {
+                        logger.info("Failed to start a Jibri session, no Jibris available")
+                        error(iq, XMPPError.Condition.service_unavailable, "no Jibri instances available")
+                    }
+                    else -> {
+                        logger.warn("Failed to start a Jibri session:${exc.message}", exc)
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt:134]: Add a space after the colon?

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+            this.jibriSession = null
+        }
+    }
+
+    /**
+     * Sends the new jibri status as presence in the conference MUC.
+     */
+    private fun publishJibriRecordingStatus(newStatus: Status, failureReason: JibriIq.FailureReason?) {
+        logger.info("Got jibri status $newStatus and failure $failureReason")
+        val jibriSession = jibriSession
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt:171]: Why are you copying it to a local here?

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    jibriDetector: JibriDetector,
+    parentLogger: Logger
+) : BaseJibri(
+    conference,
+    xmppProvider,
+    parentLogger,
+    jibriDetector
+) {
+    /**
+     * Th SIP [JibriSession]s mapped by the SIP address.
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt:55]: Typo

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+                jibriSession.start()
+                logger.info("Started Jibri session")
+                JibriIq.createResult(iq, sessionId)
+            } catch (exc: StartException) {
+                val reason = exc.message
+                logger.warn("Failed to start a Jibri session: $reason", exc)
+                sipSessions.remove(iq.sipAddress)
+                when (exc) {
+                    is AllBusy -> error(iq, XMPPError.Condition.resource_constraint, "all Jibris are busy")
+                    is NotAvailable -> error(iq, XMPPError.Condition.service_unavailable, "no Jibri instances available")
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt]: Elsewhere you changed this message to "no Jibris available" I think?

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        if (JibriIq.Status.OFF == newStatus) {
+            val sipAddress = jibriSession.sipAddress
+            sipSessions.remove(sipAddress)
+            logger.info("Removing SIP call: $sipAddress")
+        }
+    }
+
+    /**
+     * Updates status of specific [JibriSession]. Jicofo adds multiple
+     * [SipCallState] MUC presence extensions to it's presence. One for
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt:135]: Typo, not new: "its"

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        newStatus: JibriIq.Status,
+        failureReason: JibriIq.FailureReason?
+    ) {
+        val sipCallState = SipCallState().apply {
+            setState(newStatus)
+            this.failureReason = failureReason
+            sipAddress = session.sipAddress
+            sessionId = session.sessionId
+        }
+        logger.info("Publishing new state: ${session.sipAddress} ${sipCallState.toXML()} in: " + conference.roomName)
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt:152]: Use string interpolation for `conference.roomName` too.

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+            true
+        },
+        TaskPools.ioPool
+    )
+
+    /**
+     * The logger instance pass to the constructor that wil be used by this
+     * instance for logging.
+     */
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:77]: This comment isn't actually accurate given `createChildLogger`.

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+    protected val logger: Logger = parentLogger.createChildLogger(BaseJibri::class.simpleName)
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:78]: Now that we're using `Logger2` do we want to add conference-specific parameters here?

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+     * @return a list with all [JibriSession]s used by this instance.
+     */
+    abstract val jibriSessions: List<JibriSession>
+
+    /**
+     * This method will be called when start IQ arrives from Jitsi Meet
+     * participant and [.getJibriSessionForMeetIq] returns
+     * <tt>null</tt>. The implementing class should allocate and store new
+     * [JibriSession]. Once [JibriSession] is created it must be
+     * started by the implementing class.
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:110]: Articles - "a new JibriSession", "Once the JibriSession"

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+        // Process if it belongs to an active recording session
+        val session = getJibriSessionForMeetIq(iq)
+        if (session != null && session.accept(iq)) {
+            return session.processJibriIqRequestFromJibri(iq)
+        }
+        if (iq.action == Action.UNDEFINED) {
+            return error(iq, XMPPError.Condition.bad_request, "undefined action")
+        }
+
+        // verifyModeratorRole create 'not_allowed' error on when not moderator
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:177]: Grammar confused

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+                // must be that they're all busy.
+                logger.info("Failed to start a Jibri session, all Jibris were busy")
+                error(iq, XMPPError.Condition.resource_constraint, "all Jibris are busy")
+            }
+            iq.action == Action.STOP && session != null -> {
+                session.stop(iq.from)
+                IQ.createResultIQ(iq)
+            }
+            else -> {
+                logger.warn("Discarded: " + iq.toXML() + " - nothing to be done, ")
```
Comment (JonathanLennox) [inline src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:197]: Use string interpolation.

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt (hunk 16 above)
Comment (bgrozev) [reply src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:78]: I don't know that we can add anything useful, it already inherits the conference name.

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatMemberImpl.java (hunk 5 above)
Comment (bgrozev) [reply src/main/java/org/jitsi/impl/protocol/xmpp/ChatMemberImpl.java:232]: It looked suspicous to me, because we allow moderators to mute people, etc. So I thought the intention was to use "<=", and we never caught it because we have `allowners` in most places.

In the end I kept the code but removed the comment, because the new code reads clearly and because I hope all this "start muted" code is going to go away soon.

File: src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt (hunk 6 above)
Comment (bgrozev) [reply src/main/kotlin/org/jitsi/jicofo/jibri/BaseJibri.kt:69]: Because when I add the "service" XMPP connection for jibries I'll need two separate, one for each connection

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt (hunk 10 above)
Comment (bgrozev) [reply src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt:171]: Because the field might change to `null`.

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt (hunk 7 above)
Comment (bgrozev) [reply src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt:68]: I brought it back 👍

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt (hunk 14 above)
Comment (bgrozev) [reply src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt]: Removed that, it's now part of the logger context 👍

File: src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt (hunk 12 above)
Comment (bgrozev) [reply src/main/kotlin/org/jitsi/jicofo/jibri/JibriSipGateway.kt]: I think that changed in a later commit. I tried to make them all consistent now.

**Review by JonathanLennox — APPROVED** [review]
